### ISSUE TITLE
Allow for an optional step in the Wizard

### ISSUE THREAD

**Issue by gbraad** [patternfly/patternfly-design#1116]
After a lot of quick-and-dirty hacks, I was finally able to get an optional step in the `Wizard`. Only when the option for OpenShift is chosem we need to ask for the necessary pull secret. When you would choose Podman, it would not need this.

![image](https://user-images.githubusercontent.com/1894/146633820-5e24a7ac-000e-4e79-952b-7e2a76c9fa18.png)

I came up with the following:
  * define a state that control this choice `this.state.preset`
  * create the steps `2 => 'Choose your preset'`, `3 => 'Provide pull secret'` and `4 => 'Review selection'`
    * when `preset === "podman" step 3 is disabled
    * otherwise; `preset === "openshift"` step 3 is necessary
  *  step 3 has `canJumpTo: stepIdReached >= 2 && this.state.preset === "openshift"`
  * `WizardContextConsumer`
    * when `activeStep.name === "Choose your preset")` then 
    *  `Button` has specific `onClick` logic to perform the navaigation
 
```jsx
                            <Button style={{width: 120}} variant="primary" type="submit" onClick={() =>
                                {
                                    if(this.state.preset === "podman") {
                                        goToStepById(4);    // to skip to the review
                                    } else {
                                        onNext();   // perform normal behaviour to navigate to 'Provide pull secret'
                                    }
                                }}>
                            Next
                            </Button>
```

I had to resort to `activeStep.name` checks. The `id` was unreliable in use as I originally wanted to use substeps to peform this option step. As for the substep you can easily use a conditional (ternary) operator `? : ` to include a step an additional like:

```
  steps: [
     {
          name: 'Choose preset',
          component: < />
     }, 
     {this.state.preset === "openshift") ? 
     {
          name: 'Provide pull secret',
          component: < />
      } : {}
  }
```

this can not be skipped or navigated around easily. It will be removed (and therefore be very visible), but the issue is that the `id` values for a substep aren't returned?

This should be easier.

**Comment by mmenestr** [comment]
@gbraad Have you considered using a progressive wizard for this use case? A progressive wizard only shows you the steps as you reach them. So in your case, you could hide step 3 and 4, until a user has made a selection in Step 2. If in Step 2 they choose OpenShift, the next step would be "Provide pull secret", but if they choose Podman, then the next step would be "Review selection". And the user would only see the next step after clicking "Next".

Does this make sense for what you're trying to do?

**Comment by gbraad** [comment]
If I am not mistaken, the progressive does not show the total possible steps. This to me sounds like a bad indication of needed effort. Most people drop out of filling a questionnaire if they do not see progress or the end goal. I wanna prevent this unknown.

**Comment by mmenestr** [comment]
That is true, it doesn't show the next steps.I wonder if you could do a hybrid situation, where you change the steps as the user goes, instead of using this custom disablement solution.

So, you'd show all the presumed steps at all times, but if the user in step 2 chooses Podman, and therefore doesn't need to do the "provide pull secret" step, then you could hide that step as soon as they press next?

**Comment by mmenestr** [comment]
@gbraad yes, I think that would make sense^

**Comment by mcarrano** [comment]
@mcoker any reason this can't be implemented this way?

**Comment by mcoker** [comment]
@mcarrano I don't see why not, though I think a react dev should chime in on how to best do that and if we need to do any work to make it easier.

**Comment by mcarrano** [comment]
@nicolethoen @tlabaj any input here?

**Comment by tlabaj** [comment]
@mcarrano @mcoker We should be able to this from the React side.  We would probably want to add a `disabled` prop to the `WizardStep` interface.  If true, we would skip the step when next is clicked and the next step is disabled.  We would also disable the step in side nav.

**Comment by mcarrano** [comment]
Sounds good @tlabaj . I will transfer to React and get this onto the roadmap then. Thanks for this suggestion @gbraad .
